cmd/log/src/lib.rs

```diff
@@ -0,0 +1,267 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! ## `humility log`
+//!
+//! `humility log` reads and displays any Hubris string buffers (as created
+//! via the `stringbuf!` macro in the Hubris `ringbuf` crate).  e.g.:
+//!
+//! ```console
+//! % humility log -m LOG_RINGBUF
+//! humility: attached via J-Link
+//! humility: stringbuf ringbuf::stringbuf::LOG__STRINGBUF in jefe:
+//! 5532: Task #2 Illegal instruction
+//! 5535: Task #2 Memory fault at address 0x0
+//! 5537: Task #2 Illegal instruction
+//! ...
+//! ```
+//!
+//! Use `-m` or `--monitor` to continuously monitor the buffer, otherwise will just print the
+//! current log and exit.  When in monitor mode, the time (ms) between scanning for new log entries
+//! can be changed with '-d' or '--delay', it defaults to 100ms.
+//!
+//! If an argument is provided, only string buffers that have a name that
+//! contains the argument as a substring, or are in a task that contains
+//! the argument as a substring will be displayed.  
+//!
+
+use anyhow::{bail, Result};
+use clap::Command as ClapCommand;
+use clap::{CommandFactory, Parser};
+use humility::cli::Subcommand;
+use humility::core::Core;
+use humility::hubris::*;
+use humility::reflect::{self, Load, Value};
+use humility_cmd::doppel::{StaticCell, Stringbuf};
+use humility_cmd::{Archive, Attach, Command, Validate};
+use std::str;
+use std::thread;
+use std::time;
+
+#[derive(Parser, Debug)]
+#[clap(name = "stringbuf", about = env!("CARGO_PKG_DESCRIPTION"))]
+struct StringbufArgs {
+    /// list all stringbufs
+    #[clap(long, short)]
+    list: bool,
+    /// continously update the log
+    #[clap(long, short, conflicts_with = "list")]
+    monitor: bool,
+    /// how long to delay in ms between polling for new log entries
+    #[clap(
+        long,
+        short,
+        default_value = "100",
+        requires = "monitor",
+        conflicts_with = "list"
+    )]
+    delay: u64,
+    /// print only a single stringbuffer by substring of name
+    #[clap(conflicts_with = "list")]
+    name: Option<String>,
+}
+
+fn stringbuf_read(
+    hubris: &HubrisArchive,
+    core: &mut dyn Core,
+    definition: &HubrisStruct,
+    ringbuf_var: &HubrisVariable,
+    last_last: u64,
```

Comment:
I still think we could use a better name here :).

Reply:
changed to `prev_last_idx`